<a href="https://colab.research.google.com/github/apchavezr/16.-Aprendizaje-Profundo-para-Ciencia-de-Datos/blob/main/Validacion_Cruzada_CaliforniaHousing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Validación cruzada con Scikit-learn

## Objetivo
Aplicar validación cruzada con K pliegues (K-fold cross-validation) para estimar el desempeño de un modelo de regresión, utilizando el conjunto de datos California Housing.

In [1]:
# Importar bibliotecas necesarias
from sklearn.datasets import fetch_california_housing
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score
import numpy as np

## Paso 1: cargar y preparar los datos

In [2]:
# Cargar el conjunto de datos
X, y = fetch_california_housing(return_X_y=True)

## Paso 2: definir el modelo

In [3]:
# Crear el modelo de árbol de regresión
model = DecisionTreeRegressor(random_state=42)

## Paso 3: aplicar validación cruzada

In [4]:
# Validación cruzada con 5 folds, usando R² como métrica
scores = cross_val_score(model, X, y, cv=5, scoring='r2')

# Mostrar resultados por fold y promedio
print("R² por fold:", scores)
print("Promedio R²:", scores.mean())

R² por fold: [0.27093461 0.41372445 0.43912441 0.23566991 0.41875969]
Promedio R²: 0.355642615410327


## Interpretación
* El modelo tiene rendimiento consistente entre pliegues, lo que sugiere estabilidad.

* Un promedio R² de 0.3556 indica que hay espacio para mejorar el modelo (por ejemplo, ajustando hiperparámetros, probando otros algoritmos o incorporando ingeniería de características).

* La validación cruzada ha permitido estimar el rendimiento sin depender de una única división entrenamiento/prueba, lo que reduce el sesgo de evaluación.

## Recomendaciones
- Para mayor estabilidad, se puede aumentar el número de folds (`cv=10`).
- Utilice otras métricas (`accuracy`, `f1`, `neg_mean_squared_error`, etc.) dependiendo del tipo de problema.
- Esta técnica es útil tanto en regresión como en clasificación.